### PR TITLE
Record: 7-gram N-gram Cache (0.8960 bpb)

### PR DESCRIPTION
**Pull request by armantsaturian**
## 7-gram N-gram Cache + XSA-All + LeakyReLU² + Parallel Muon

**val_bpb: 0.8960** (3-seed mean, std 0.0004) | **~15.92 MB** | 8×H100 SXM

### Results

| Seed | Pre-cache bpb | **Post-cache bpb** | Cache gain | Eval time | Artifact |
|------|---------------|-------------------|------------|-----------|----------|
| 1337 | 1.1488 | **0.8956** | -0.2532 | 136s | 15,917,799 |
| 42 | 1.1488 | **0.8959** | -0.2529 | 136s | 15,921,427 |
| 2025 | 1.1500 | **0.8966** | -0.2534 | 138s | 15,911,611 |
| **Mean** | **1.1492** | **0.8960** | **-0.2532** | **~137s** | |

### Approach: Streaming 7-gram N-gram Cache at Eval

A single-pass streaming n-gram cache built during sliding window evaluation. For each scored token:

1. **Score** neural model NLL under `torch.inference_mode()`
2. **Cache lookup** across n-gram orders 2-7, with near-zero backoff beta (1e-6) — longest match dominates
3. **Blend** with fixed 80/20 cache/neural mix (alpha=0.20)
4. **Update** cache with scored token's n-grams (score-before-update, single-pass)

No TTT, no online mixer — the cache alone delivers **-0.2532 bpb** over the base model.

### Legality

- ✅ Single-pass, score-before-update (cache updated only after scoring)
- ✅ All eval under `torch.inference_mode()` (no gradients, no weight mutation)
- ✅ No training on validation data
- ✅ No multi-pass replay or phase-2 re-scoring
- ✅ No external downloads — fully self-contained

### Timing

| Phase | Time |
|-------|------|
| Training | 600s (10 min cap) |
| Eval (sliding window + 7-gram cache) | ~137s |
| **Total** | **~737s** |

### Built on

- [PR #549](https://github.com/openai/parameter-golf/pull/549) — LeakyReLU² + Legal TTT + Parallel Muon (base model)
- [PR #741](https://github.com/openai/parameter-golf/pull/741) — N-gram cache inspiration
- [PR #745](https://github.com/openai/parameter-golf/pull/745) — XSA-all, Hedge mixer concepts
- PAQ/cmix compression literature — backoff and blending techniques

See [full README](records/track_10min_16mb/2026-03-26_7gram_NgramCache_0.8960/README.md) for ablation, architecture details, and what we tried and discarded.

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — 7-gram Hashed N-gram Cache

**BPB:** 0.8960 (3-seed, std 0.0004) | **Artifact:** ~15.92 MB | **Vocab:** SP1024 | **Compliance:** FLAG — same family as PR #779

**What this does:** A streaming 2-to-7-gram cache built during sliding-window eval. For each scored position, the cache looks up two hashed counts — a context-only key and a "full" key that includes the target token — then blends the resulting estimate with the neural probability at `alpha=0.20`.

**What I found in the code (SHA `7df5136`):**

The cache keys are built in `ngram_hash_keys` at [`records/track_10min_16mb/2026-03-26_7gram_NgramCache_0.8960/train_gpt.py` lines 129–136](https://github.com/openai/parameter-golf/pull/797/files):

```python
def ngram_hash_keys(tokens_np, pos, order, mask, primes):
    ctx = order - 1
    h = np.zeros(pos.size, dtype=np.uint64)
    for k in range(ctx):
        h ^= tokens_np[pos - (ctx - k)].astype(np.uint64) * primes[k % 5]
    ctx_idx = (h & mask).astype(np.int64)
    full_idx = ((h ^ (tokens_np[pos].astype(np.uint64) * primes[ctx % 5])) & mask).astype(np.int64)
    return ctx_idx, full_idx
```

`full_idx` hashes the **target token** `tokens_np[pos]` into the lookup key. At the scoring call site (line 1146), `fk` is passed into `ngram_full[n][fk]` and combined with the context count at line 1151:

```python
cache_prob[vi] = np.clip(
    (np.minimum(fc, cc) + args.ngram_backoff_beta * prev) /
    (cc + args.ngram_backoff_beta),
    0.0, 1.0,
)
```

So the cache is queried once per `(context, target)` pair via a key that embeds the target, and the returned count drives the blended probability for that target.

**Questions / flags:**

- This is the same `full_key = ((ctx_hash ^ (target * primes[k])) & mask)` construction that @valerio-oai disallowed on [PR #779](https://github.com/openai/parameter-golf/pull/779#issuecomment-4145781641) (2026-03-27): hashed n-gram caches that fold the scored token into the lookup key are not compliant because `p_t` effectively depends on `x_t` through the hash.
- Per Issue #1017 Condition 1, `p_t` may depend only on the artifact and `x_1...x_{t-1}`. Because `full_idx` is a function of `x_t`, the lookup `ngram_full[n][full_idx]` cannot be computed from the prefix alone — the probability assigned to each candidate token would require hashing with that candidate, which is precisely the pattern the earlier ruling marked as invalid.
- The PR body reports a **pre-cache BPB of 1.1492** and a post-cache BPB of **0.8960**, a gain of −0.2532. The entire record-track delta (the sub-1.15 → sub-0.90 jump) appears to come from the disallowed cache, not from the base model.
- Is there a context-only variant that reproduces using **only** `ctx_idx` (which is a function of `x_1...x_{t-1}` alone) and computes per-candidate probabilities over the full vocab from that single context row? That would be compliant in the spirit of #779/#1017.

**Verdict:** COMPLIANCE FLAG — hashed-target n-gram cache, same family as PR #779.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** CLOSE as non-compliant, pending author revision to a context-only cache. The submission is clean on single-pass / score-before-update / no grads, but the cache key construction at line 135 is the pattern that was explicitly ruled out on PR #779.

---

### Update 2026-04-11 — Gauntlet rerun on CT2038

The first version of this review noted the CPU gauntlet was skipped due to script size + FA3 dependency on the local workstation, with the framing "compliance audit was the blocking concern." Provisioned a dedicated container (`proteus-engine`, 128 GB RAM, 32 cores) with Triton 3.6.0 + a flash_attn stub and re-ran the gauntlet from scratch on `train_gpt.py` at SHA `7df5136`:

```
[PASS] Import (0.0s)
[PASS] Hyperparameters: dim=512, layers=11, heads=8, vocab=1024
[PASS] Model: 26,993,756 params (0.7s)
[PASS] Forward pass: loss=6.9366 (0.0s)
[INFO] Code size: 108,267 bytes (105.7 KB)
[PASS] Artifact: 4,629,887 bytes (28.9% of 16MB) [int6+lzma] (17.5s)
Result: ✓ ALL PASS
```

`forward_loss=6.9366`, `total_params=26,993,756`. The base neural model imports and runs cleanly — the compliance flag remains entirely about the n-gram lookup key construction at line 135 of `ngram_hash_keys`. Verdict and recommendation are unchanged: CLOSE under the @valerio-oai PR #779 ruling. The base PR #549-style stack at the pre-cache 1.1492 BPB number is unaffected and would be welcome in a context-only resubmission.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). CPU gauntlet (CT2038 proteus-engine, 2026-04-11): all 10 checks PASS — import, hyperparams, model creation (26.99M params), forward pass (loss 6.9366), code size 108,267 B, artifact 4.63 MB int6+lzma (28.9% of 16 MB budget). The compliance flag is solely about the target-in-key n-gram lookup, not the base neural code. AI tooling: review drafted with Claude Code (Sonnet/Opus) using an internal review template; all citations, file paths, and compliance audits were verified against the PR's actual code at SHA `7df51367caa75559f1b00c51e72b27ad7dfafe42`.*
